lib/salus/scanners/osv/maven_osv.rb

```diff
@@ -0,0 +1,150 @@
+        report_error(err_msg)
+        return
+      end
+
+      @osv_vulnerabilities ||= fetch_vulnerabilities(MAVEN_OSV_ADVISORY_URL)
+      if @osv_vulnerabilities.nil?
+        err_msg = "No vulnerabilities found to compare."
+        bugsnag_notify("MavenOSV: #{err_msg}")
+        return report_error("MavenOSV: #{err_msg}")
+      end
+
+      # Fetch vulnerable dependencies.
+      # Dedupe and select Github Advisory over other sources if available.
+      results = []
+      grouped = match_vulnerable_dependencies(dependencies).group_by { |d| d[:ID] }
+      grouped.each do |_key, values|
+        vuln = {}
+        values.each do |v|
+          vuln = v if v[:Database] == GITHUB_DATABASE_STRING
+        end
+        results.append(vuln.empty? ? values[0] : vuln)
+      end
+
+      # Report scanner status
+      return report_success if results.empty?
+
+      report_failure
+      log(JSON.pretty_generate(results))
+    end
+
+    # Match if dependency version found is in the range of
+    # vulnerable dependency found.
+    def version_matching(version, version_ranges)
+      vulnerable_flag = false
+      version_found = SemVersion.new(version)
+      # If version range length is 1, then no fix available.
+      if version_ranges.length == 1
+        introduced = SemVersion.new(
+          version_ranges[0]["introduced"]
+        )
+        vulnerable_flag = true if version_found >= introduced
+      # If version range length is 2, then both introduced and fixed are available.
+      elsif version_ranges.length == 2
+        introduced = SemVersion.new(
+          version_ranges[0]["introduced"]
+        )
+        fixed = SemVersion.new(
+          version_ranges[1]["fixed"]
+        )
+        vulnerable_flag = true if version_found >= introduced && version_found < fixed
+      end
+
+      vulnerable_flag
+    end
+
+    # Compare vulnerabilities found with dependencies found
+    # and return vulnerable dependencies
+    def match_vulnerable_dependencies(dependencies)
+      results = []
+      dependencies.each do |dependency|
+        lib = "#{dependency['group_id']}:#{dependency['artifact_id']}"
+
+        unless dependency['version'].nil?
+          version = dependency['version']
+          # If version is of the format '${deps.version}', log it.
+          if version.match(/\${(.*)\.version}/)
+            bugsnag_notify("MavenOSV: Found #{lib}:#{version} with incompatible format.")
+            break
+          end
+
+          package_matches = @osv_vulnerabilities.select do |v|
+            v.dig("package", "name") == lib
+          end
+          package_matches.each do |m|
+            m["ranges"].each do |version_ranges|
+              if version_ranges["type"] == "SEMVER" || version_ranges["type"] == "ECOSYSTEM"
+                introduced = version_ranges["events"][0]["introduced"]
+                fixed = if version_ranges["events"].length == 2
+                          version_ranges["events"][1]["fixed"]
```

Comment:
do `introduced` and `fixed` always exist as keys?

Reply:
introduced always exist, `fixed` doesn't that's why the check for length == 2. So both `introduced` and `fixed` will exist.

Reply:
But is it possible for length to be 2, but the keys are not introduced/keys?

Reply:
Could be (for example - if the input we get is bad) but I haven't seen an example of it since they are following the OSV Format. So according the format these fields would need to be there.